crates/node/engine/src/task_queue/tasks/synchronize/task.rs

```
@@ -0,0 +1,137 @@
+use async_trait::async_trait;
+use kona_genesis::RollupConfig;
+use op_alloy_provider::ext::engine::OpEngineApi;
+use std::sync::Arc;
+use tokio::time::Instant;
+
+/// The [`SynchronizeTask`] executes an `engine_forkchoiceUpdated` call with the current
+/// [`EngineState`]'s forkchoice, and no payload attributes.
+#[derive(Debug, Clone)]
+pub struct SynchronizeTask {
+    /// The engine client.
+    pub client: Arc<EngineClient>,
+    /// The rollup config.
+    pub rollup: Arc<RollupConfig>,
+    /// The sync state update to apply to the engine state.
+    pub state_update: EngineSyncStateUpdate,
+}
+
+impl SynchronizeTask {
+    /// Creates a new [`SynchronizeTask`].
+    pub const fn new(
+        client: Arc<EngineClient>,
+        rollup: Arc<RollupConfig>,
+        state_update: EngineSyncStateUpdate,
+    ) -> Self {
+        Self { client, rollup, state_update }
+    }
+
+    /// Checks the response of the `engine_forkchoiceUpdated` call, and updates the sync status if
+    /// necessary.
+    fn check_forkchoice_updated_status(
+        &self,
+        state: &mut EngineState,
+        status: &PayloadStatusEnum,
+    ) -> Result<(), SynchronizeTaskError> {
+        match status {
+            PayloadStatusEnum::Valid => {
+                if !state.el_sync_finished {
+                    info!(
+                        target: "engine",
+                        "Finished execution layer sync."
+                    );
+                    state.el_sync_finished = true;
+                }
+
+                Ok(())
+            }
+            PayloadStatusEnum::Syncing => {
+                // If we're not building a new payload, we're driving EL sync.
+                warn!(target: "engine", "Attempting to update forkchoice state while EL syncing");
+                Ok(())
+            }
+            s => {
+                // Other codes are not expected.
+                Err(SynchronizeTaskError::UnexpectedPayloadStatus(s.clone()))
+            }
```

I _think_ in this case we should return a Critical error. The `forkchoice_updated` method shouldn't _ever_ return the other status enum if we're not providing payloads

---

It should only critical if the error is actually invalid. Accepted is fine. We should be very explicit about what is a critical error

---

Although I am also a bit unsure about returning a critical error here because we might also get affected if there is any change to the downstream api (if `alloy` ever decide to return `Accepted` inside `forkchoice_update`)